std/assembly/internal/array.ts

```diff
@@ -10,7 +10,59 @@ import {
 /** Obtains the default comparator for the specified type. */
 @inline
 export function defaultComparator<T>(): (a: T, b: T) => i32 {
-  return (a: T, b: T): i32 => (<i32>(a > b) - <i32>(a < b)); // compiles to a constant table index
+  if (isInteger<T>() && isSigned<T>()) {
```

Comment:
Does this work with tree-shaking? Afaik each of these must be a standalone condition for it to work because there'll otherwise be a temp. local (and yeah, that still needs to be optimized).

Reply:
I don't know about this limit. Thanks. Fixed!

Reply:
Just tried it out with

```ts
function foo<T>(a: T, b: T): T {
  if (isInteger<T>() && isSigned<T>()) {
    return a - b;
  } else {
    return a + b;
  }
}

foo<i32>(1, 2);
foo<u32>(1, 2);
foo<f32>(1, 2);
```

and it actually somehow works.

```wat
 (func $tests/compiler/empty/foo<i32> (; 0 ;) (type $iii) (param $0 i32) (param $1 i32) (result i32)
  (return
   (i32.sub
    (get_local $0)
    (get_local $1)
   )
  )
 )
 (func $tests/compiler/empty/foo<u32> (; 1 ;) (type $iii) (param $0 i32) (param $1 i32) (result i32)
  (return
   (i32.add
    (get_local $0)
    (get_local $1)
   )
  )
 )
 (func $tests/compiler/empty/foo<f32> (; 2 ;) (type $fff) (param $0 f32) (param $1 f32) (result f32)
  (return
   (f32.add
    (get_local $0)
    (get_local $1)
   )
  )
 )
```

Guess Binaryen is able to precompute that nonetheless, sorry :)

Reply:
Cool!